thrall/app/lib/ThrallStreamProcessor.scala

```diff
@@ -1,62 +1,101 @@
 package lib
 
+import java.time.Instant
+
 import akka.actor.ActorSystem
-import akka.stream.scaladsl.{GraphDSL, MergePreferred, Source}
+import akka.stream.scaladsl.{GraphDSL, MergePreferred, MergePrioritized, Source}
 import akka.stream.{Materializer, SourceShape}
 import akka.{Done, NotUsed}
 import com.contxt.kinesis.KinesisRecord
 import com.gu.mediaservice.lib.DateTimeUtils
 import com.gu.mediaservice.lib.aws.UpdateMessage
 import com.gu.mediaservice.lib.logging._
 import com.gu.mediaservice.model.ThrallMessage
-import lib.kinesis.{MessageTranslator, ThrallEventConsumer}
-
+import lib.kinesis.{MessageTranslator,ThrallEventConsumer}
 import scala.concurrent.Future
 import scala.util.{Failure, Success}
 
 sealed trait Priority
-case object LowPriority extends Priority {
+case object AutomationPriority extends Priority {
   override def toString = "low"
 }
-case object HighPriority extends Priority {
+case object ReingestionPriority extends Priority {
+  override def toString = "lowest"
+}
+case object UiPriority extends Priority {
   override def toString = "high"
 }
-case class TaggedRecord(record: KinesisRecord, priority: Priority) extends LogMarker {
-  override def markerContents = Map(
+
+case class TaggedRecord[T](payload: T,
+                           arrivalTimestamp: Instant,
+                           priority: Priority,
+                           markProcessed: () => Unit) extends LogMarker {
```

Comment:
I like this. Just to check my understanding, I'm guessing we're asking for these properties explicitly, rather than wrapping a `Record`, b/c [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) or similar? I'm torn as to whether a worth a comment to describe what a TaggedRecord is, or why it exists, might be nice.

Reply:
It's not just because of the 'principal of least knowledge', but also because `TaggedRecord` now represents not only messages coming from Kinesis but also our new messages that originate within thrall. @

Reply:
code comment added in https://github.com/guardian/grid/pull/3271/commits/b44df43a3f8aa4cdf335e86c2c7b8ceff330f431